globus_automate_client/cli/callbacks.py

```diff
@@ -85,15 +85,21 @@ def text_validator_callback(message: str) -> str:
     return message
 
 
-def principal_validator(principals: List[str]) -> List[str]:
+def _base_principal_validator(
+    principals: List[str], *, special_vals=set()
```

Comment:
I think using `set()` here is safe in terms of re-use. I guess something like: `special_vals: Set[str] = frozenset()` would be ultimately safe?

I'm just comparing this to general guidance where default values on kwargs like `[]` or `{}` for lists or dicts are not considered safe because, if I recall correctly, those values get created at interpret time and could get changed on multiple calls to the function.

Reply:
That makes sense, I've convinced myself that using mutable values is fine bc they're used in the context of a CLI command but I think best practices should prevail. I think setting the default to `None` would be best in this case.

Reply:
I like frozenset. I've been using `Optional[CollectionType[base type]] = None` for so long, but recently occurred to me that there are cases where `Iterable[base type] = ()` for an empty tuple which is immutable or `Set[base type] = frozenset()` for sets is doable.

This isn't a silver bullet of course. `Iterable` isn't always sufficient (but maybe catches the old string is iterable problem when specified this way (the values returned from iterating a string aren't also strings that happen to be one character long, are they?)). And, I don't believe there's any form of a dict which is immutable, so there'll still probably be things that are `[Optinonal[Dict[...]] = None`, but maybe we apply this sort of thing when we can.

And, even in cases like this which are practically thinking probably safe, we start to train ourselves and get our code having some bit of consistency.

Reply:
Consistency is a big win and probably worth re-wiring some of the assumptions that we bring into our code. I like the `Set[base type] = frozenset()` solution with maybe one change - `AbstractSet[str] = frozenset()`. The change has the benefit of silencing mypy warnings about whether a set or frozenset is used when the func gets called.